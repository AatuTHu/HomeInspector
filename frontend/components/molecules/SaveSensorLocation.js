import { useState } from 'react'
import variables from '../../env'
import InfoBox from '../atoms/InfoBox'
import InputBox from '../atoms/InputBox'
import Buttons from '../atoms/Buttons'
import { View, Text } from 'react-native'
import { styles } from "../../styles/styles"

export default function saveSensorLocation({ location, locationURL}) {

    const { apiKey } = variables
    const [newLocation, setNewLocation] = useState('');
    const [ statusText, setStatusText ] = useState('');

    const saveLocation = async() => {
        const response = await fetch(locationURL, {
           method: "POST",
           headers: { "Content-Type": "application/x-www-form-urlencoded" },
           body: `apiKey=${apiKey}&location=${newLocation}`,
         })
         
         if(response.status === 202) {
           setStatusText("Location saved")
         } else if (response.status === 403) {
           setStatusText("There was an error, Try again later")
         }
       }

  return (<>
    <InfoBox text = "Device location" textStyle = 'headingText' type = 'heading'/>  
        <InputBox placeholder= "where ..." text = { location } newText = { setNewLocation } onSubmitEditing={ saveLocation }/>
    <Buttons text = "Save" type = "save" onPress={ saveLocation } />
    <View style = { styles.block}>
        <Text style = { styles.statusText }>{statusText}</Text>
    </View>
  </>)
}