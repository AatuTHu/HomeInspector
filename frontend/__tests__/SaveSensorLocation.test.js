import React from 'react';
import SaveSensorLocation from '../components/molecules/SaveSensorLocation';
import { render, fireEvent, waitFor } from '@testing-library/react-native';

describe('SaveSensorLocation', () => {

afterEach(() => {
  jest.clearAllMocks();
})

it('renders correctly', () => {
  const screen = render(<SaveSensorLocation/>);
  expect(screen).toMatchSnapshot();
});

  

it('should set statusText to "Location saved" when response status is 202', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        status: 202,
        json: () => Promise.resolve({}),
      })
    );
    const { getByText } = render(<SaveSensorLocation setCurrentTempLoc={jest.fn()} setCurrentHumLoc={jest.fn()} selected = "humidity"/>);
    const saveButton = getByText('Save');
    fireEvent.press(saveButton);
    await waitFor(() => expect(getByText('Location saved')).toBeTruthy());
})


it('should set statusText to "There was an error, Try again later" when response status is 403', async () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 403,
      json: () => Promise.resolve({}),
    })
  );

  const { getByText } = render(<SaveSensorLocation setCurrentTempLoc={jest.fn()} setCurrentHumLoc={jest.fn()} selected = "humidity"/>);
  const saveButton = getByText('Save');
  fireEvent.press(saveButton);
  await waitFor(() => expect(getByText('There was an error, Try again later')).toBeTruthy());
})
})