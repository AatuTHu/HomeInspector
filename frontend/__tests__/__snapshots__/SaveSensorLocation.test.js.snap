// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveSensorLocation renders correctly 1`] = `
[
  <View
    style={
      {
        "borderRadius": 22,
        "height": 50,
        "justifyContent": "center",
        "marginTop": 20,
        "width": "95%",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 20,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Device location
    </Text>
  </View>,
  <View
    style={
      {
        "backgroundColor": "#E8F1F2",
        "height": 50,
        "justifyContent": "center",
        "marginTop": 10,
        "width": "95%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="where ..."
      style={
        {
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#00A352",
        "borderRadius": 20,
        "borderWidth": 1,
        "marginTop": 20,
        "opacity": 1,
        "padding": 10,
        "width": "95%",
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      Save
    </Text>
  </View>,
  <View
    style={
      {
        "height": 100,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "darkgreen",
          "fontSize": 20,
          "justifyContent": "center",
          "padding": 10,
          "textAlign": "center",
        }
      }
    />
  </View>,
]
`;
